File: agent/tools/marketing/twitter.py
"""Twitter (X) tools — post and read on behalf of the founder.

Auth priority: IntegrationConnection (toolkit='twitter', status='active')
→ TWITTER_BEARER_TOKEN env. Tokens are OAuth 2.0 user-context bearer
tokens — the same shape works for read + write on v2 endpoints.

All tools return strings (URLs or error descriptions). Wave 2 wraps the
mutating tool with the approval gate at the manager layer.
"""

from __future__ import annotations

import os
from typing import Any

import httpx
import structlog

from agents import function_tool
from db import supabase
from tools._approval import gate_or_execute

logger = structlog.get_logger(__name__)

_TWITTER_API = "https://api.twitter.com/2"
_TIMEOUT = 20.0


async def _get_twitter_token(space_id: str) -> str | None:
    if space_id:
        try:
            db = await supabase()
            res = await (
                db.table("IntegrationConnection")
                .select("accessToken")
                .eq("spaceId", space_id)
                .eq("toolkit", "twitter")
                .eq("status", "active")
                .maybe_single()
                .execute()
            )
            if res.data and res.data.get("accessToken"):
                return res.data["accessToken"]
        except Exception as err:  # noqa: BLE001
            logger.warning("twitter_token_lookup_failed", space_id=space_id, error=str(err)[:200])

    return os.environ.get("TWITTER_BEARER_TOKEN")


def _auth_headers(token: str) -> dict[str, str]:
    return {
        "Authorization": f"Bearer {token}",
        "Content-Type": "application/json",
    }


def _handle_response(resp: httpx.Response, action: str) -> str | None:
    if resp.status_code == 401:
        return f"Twitter auth failed for {action}. Reconnect X in settings or refresh TWITTER_BEARER_TOKEN."
    if resp.status_code == 403:
        return f"Twitter permission denied for {action}. Token may lack required scopes (tweet.write/users.read)."
    if resp.status_code == 429:
        return f"Twitter rate limit hit for {action}. Retry after cooldown."
    if not resp.is_success:
        body: dict[str, Any] = {}
        try:
            body = resp.json()
        except Exception:  # noqa: BLE001
            pass
        msg = body.get("detail") or body.get("title") or resp.text[:200]
        return f"Twitter API error {resp.status_code} for {action}: {msg}"
    return None


async def _get_me(client: httpx.AsyncClient, token: str) -> dict[str, Any] | str:
    """Return the authenticated user's profile dict, or an error string."""
    resp = await client.get(
        f"{_TWITTER_API}/users/me",
        headers=_auth_headers(token),
        params={"user.fields": "public_metrics,username"},
    )
    err = _handle_response(resp, "get_me")
    if err:
        return err
    return resp.json().get("data") or {}


@function_tool
async def twitter_post_tweet(
    text: str,
    reply_to_tweet_id: str = "",
    space_id: str = "",
) -> str:
    """Post a tweet. Returns the tweet URL on success.

    Optionally reply to an existing tweet by passing reply_to_tweet_id.
    """

    async def _execute() -> str:
        token = await _get_twitter_token(space_id)
        if not token:
            return "No Twitter token found. Connect X in settings or set TWITTER_BEARER_TOKEN."

        payload: dict[str, Any] = {"text": text}
        if reply_to_tweet_id:
            payload["reply"] = {"in_reply_to_tweet_id": reply_to_tweet_id}

        async with httpx.AsyncClient(timeout=_TIMEOUT) as client:
            resp = await client.post(
                f"{_TWITTER_API}/tweets",
                headers=_auth_headers(token),
                json=payload,
            )

        err = _handle_response(resp, "post_tweet")
        if err:
            return err

        data = (resp.json() or {}).get("data") or {}
        tweet_id = data.get("id", "")
        url = f"https://x.com/i/web/status/{tweet_id}" if tweet_id else ""
        logger.info("twitter_tweet_posted", space_id=space_id, tweet_id=tweet_id)
        return f"Tweeted: {url}" if url else "Tweet posted but no id returned."

    return await gate_or_execute(
        space_id=space_id,
        department="marketing",
        action_label=f"Post tweet: {text[:80]}",
        risk="high",
        execute_fn=_execute,
    )


@function_tool
async def twitter_get_recent_tweets(limit: int = 10, space_id: str = "") -> str:
    """List recent tweets authored by the connected account."""
    token = await _get_twitter_token(space_id)
    if not token:
        return "No Twitter token found. Connect X in settings or set TWITTER_BEARER_TOKEN."

    limit = max(5, min(limit, 100))  # X API requires 5-100

    async with httpx.AsyncClient(timeout=_TIMEOUT) as client:
        me = await _get_me(client, token)
        if isinstance(me, str):
            return me
        user_id = me.get("id")
        if not user_id:
            return "Twitter: could not resolve authenticated user id."

        resp = await client.get(
            f"{_TWITTER_API}/users/{user_id}/tweets",
            headers=_auth_headers(token),
            params={"max_results": limit, "tweet.fields": "created_at"},
        )

    err = _handle_response(resp, "get_recent_tweets")
    if err:
        return err

    tweets = (resp.json() or {}).get("data") or []
    if not tweets:
        return "No recent tweets."
    return "\n".join(
        f"{t.get('id')} [{t.get('created_at', '')}]: {t.get('text', '')[:200]}"
        for t in tweets
    )


@function_tool
async def twitter_get_account_metrics(space_id: str = "") -> str:
    """Return follower/following/tweet counts for the connected account."""
    token = await _get_twitter_token(space_id)
    if not token:
        return "No Twitter token found. Connect X in settings or set TWITTER_BEARER_TOKEN."

    async with httpx.AsyncClient(timeout=_TIMEOUT) as client:
        me = await _get_me(client, token)
        if isinstance(me, str):
            return me

    metrics = me.get("public_metrics") or {}
    handle = me.get("username", "")
    return (
        f"@{handle}\n"
        f"Followers: {metrics.get('followers_count', 0)}\n"
        f"Following: {metrics.get('following_count', 0)}\n"
        f"Tweets: {metrics.get('tweet_count', 0)}"
    )

File: lib/integrations/adapters/twitter.ts
/**
 * Thin Twitter (X) v2 API adapter.
 *
 * Auth: IntegrationConnection (toolkit='twitter', status='active').
 * accessToken → OAuth 2.0 user-context bearer. Env fallback for dev/CI:
 * TWITTER_BEARER_TOKEN. Throws when neither is set.
 *
 * Read + write endpoints both use the same bearer header. We do not run
 * an OAuth dance here — that's handled upstream by the connect flow.
 */

import { supabase } from '@/lib/supabase';

const X_API = 'https://api.twitter.com/2';

async function getTwitterToken(spaceId: string): Promise<string> {
  const { data } = await supabase
    .from('IntegrationConnection')
    .select('accessToken')
    .eq('spaceId', spaceId)
    .eq('toolkit', 'twitter')
    .eq('status', 'active')
    .maybeSingle();

  const token = (data as { accessToken?: string } | null)?.accessToken;
  if (token) return token;

  const env = process.env.TWITTER_BEARER_TOKEN;
  if (env) return env;

  throw new Error(
    `Twitter: no access token found for space ${spaceId}. ` +
      `Connect X in Settings → Integrations, or set TWITTER_BEARER_TOKEN.`,
  );
}

async function xFetch(
  token: string,
  path: string,
  options?: RequestInit,
): Promise<Response> {
  return fetch(`${X_API}${path}`, {
    ...options,
    headers: {
      Authorization: `Bearer ${token}`,
      'Content-Type': 'application/json',
      ...(options?.headers ?? {}),
    },
  });
}

async function readOrThrow(res: Response, action: string): Promise<unknown> {
  if (!res.ok) {
    let body: { detail?: string; title?: string } = {};
    try {
      body = (await res.json()) as { detail?: string; title?: string };
    } catch {
      // ignore
    }
    const msg = body.detail ?? body.title ?? (await res.text().catch(() => ''));
    throw new Error(`Twitter ${action} failed: ${res.status} — ${msg}`);
  }
  return res.json();
}

interface MeResponse {
  data?: {
    id: string;
    username: string;
    public_metrics?: {
      followers_count?: number;
      following_count?: number;
      tweet_count?: number;
    };
  };
}

async function getMe(token: string): Promise<MeResponse['data']> {
  const res = await xFetch(token, '/users/me?user.fields=public_metrics,username');
  const data = (await readOrThrow(res, 'getMe')) as MeResponse;
  if (!data.data?.id) {
    throw new Error('Twitter getMe failed: response missing user id');
  }
  return data.data;
}

// ── Types ────────────────────────────────────────────────────────────────────

export interface TweetResult {
  id: string;
  url: string;
}

export interface RecentTweet {
  id: string;
  text: string;
  createdAt: string;
}

export interface AccountMetrics {
  followers: number;
  following: number;
  tweetCount: number;
  handle: string;
}

// ── Operations ───────────────────────────────────────────────────────────────

export async function twitterPostTweet(
  spaceId: string,
  opts: { text: string; replyToTweetId?: string },
): Promise<TweetResult> {
  const token = await getTwitterToken(spaceId);
  const body: Record<string, unknown> = { text: opts.text };
  if (opts.replyToTweetId) {
    body.reply = { in_reply_to_tweet_id: opts.replyToTweetId };
  }

  const res = await xFetch(token, '/tweets', {
    method: 'POST',
    body: JSON.stringify(body),
  });
  const data = (await readOrThrow(res, 'postTweet')) as { data?: { id: string } };
  const id = data.data?.id ?? '';
  return { id, url: id ? `https://x.com/i/web/status/${id}` : '' };
}

export async function twitterGetRecentTweets(
  spaceId: string,
  opts?: { limit?: number },
): Promise<{ tweets: RecentTweet[] }> {
  const token = await getTwitterToken(spaceId);
  // X API requires 5..100 for max_results.
  const limit = Math.min(Math.max(opts?.limit ?? 10, 5), 100);

  const me = await getMe(token);
  const res = await xFetch(
    token,
    `/users/${me!.id}/tweets?max_results=${limit}&tweet.fields=created_at`,
  );
  const data = (await readOrThrow(res, 'getRecentTweets')) as {
    data?: Array<{ id: string; text: string; created_at?: string }>;
  };
  const tweets = (data.data ?? []).map((t) => ({
    id: t.id,
    text: t.text,
    createdAt: t.created_at ?? '',
  }));
  return { tweets };
}

export async function twitterGetAccountMetrics(
  spaceId: string,
): Promise<AccountMetrics> {
  const token = await getTwitterToken(spaceId);
  const me = await getMe(token);
  const m = me!.public_metrics ?? {};
  return {
    followers: m.followers_count ?? 0,
    following: m.following_count ?? 0,
    tweetCount: m.tweet_count ?? 0,
    handle: me!.username,
  };
}
